File: BaseLibrary/communication/command/ACommand.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace BaseLibrary.communication.command {
    public static class ACommandExtensions {
        public static async Task SendAsync(this ACommand.Sendable instance, NetworkStream s) {
            if (instance == null) {
                throw new ArgumentNullException("Instance can not be null.");
            }
            await s.WriteLineAsync(instance.Serialize());
        }

        public static string ToString(this ACommand.Sendable instance) {
            if (instance == null) {
                throw new ArgumentNullException("Instance can not be null.");
            }
            return instance.Serialize();
        }
    }

    public abstract class ACommand {
        /// <summary>
        /// If someone implements ACommand.Sendable it have to be child of ACommand
        /// </summary>
        public interface Sendable {
            string Serialize();
        }

        /// <summary>
        /// true if request do not have valid data.
        /// </summary>
        protected bool pending = true;

        /// <summary>
        /// For send mod extensions.
        /// </summary>
        public Object[] MORE { get; protected set; }

        protected ACommand() {
            
        }

        public override string ToString() {
            ACommand.Sendable sendable = this as ACommand.Sendable;
            if (sendable != null) {
                return sendable.Serialize();
            } else {
                StringBuilder sb = new StringBuilder();
                sb.Append(base.ToString());
                sb.Append("[");
                foreach (var propertyInfo in this.GetType().GetProperties()) {
                    sb.Append(propertyInfo.Name);
                    sb.Append(":");
                    sb.Append(propertyInfo.GetValue(this, null));
                    sb.Append(";");
                }

                foreach (var fieldInfo in this.GetType().GetFields()) {
                    sb.Append(fieldInfo.Name);
                    sb.Append(":");
                    sb.Append(fieldInfo.GetValue(this));
                    sb.Append(";");
                }
                sb.Append("]");

                return sb.ToString();
            }
        }
        protected string[] SerializeMore(IEnumerable<ISubCommandFactory> subCommandFactories) {
            string[] serializedMore = new string[MORE.Length];
            for (int i = 0; i < MORE.Length; i++) {
                foreach (var factory in subCommandFactories) {
                    if (factory.Serialize(MORE[i], out serializedMore[i])) {
                        break;
                    }

                }
            }
            return serializedMore;
        }

        protected void DeserializeMore(string[] serializedMore, object[] more, IEnumerable<ISubCommandFactory> subCommandFactories) {
            Console.WriteLine($"{GetType().Name} Deserialize - more.Length:{more.Length}");
            foreach (var moreString in serializedMore) {
                foreach (var subCommandFactory in subCommandFactories) {
                    if (subCommandFactory.Deserialize(moreString, more)) {
                        break;
                    }
                }
            }
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/communication/command/ACommand.java
package BaseLibrary.communication.command;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import BaseLibrary.communication.protocol.AProtocol;
import BaseLibrary.communication.protocol.AProtocol.Array;
import BaseLibrary.communication.protocol.IInnerObject;

public abstract class ACommand {
	
	/**
	 * If someone implements ACommand.Sendable it have to be child of ACommand
	 */
	public static interface Sendable {
		String Serialize();
	}

	/**
	 * true if request do not have valid data.
	 */
	protected boolean pending = true;

	/**
	 * For send mod extensions.
	 */
	public Object[] MORE;

	/**
	 * for storing mod factories
	 */
	private static final HashMap<Class<?>, List<ISubCommandFactory>> SUB_COMMAND_FACTORIES = new HashMap<>();

	public static <T> int RegisterSubCommandFactory(ISubCommandFactory subCommandFactory, Class<T> forCommandClass) {
		int position = SUB_COMMAND_FACTORIES.size();
		List<ISubCommandFactory> factories = SUB_COMMAND_FACTORIES.get(forCommandClass);
		if (factories == null) {
			factories = new ArrayList<>();
		}
		factories.add(subCommandFactory);
		SUB_COMMAND_FACTORIES.put(forCommandClass, factories);
		return position;
	}

	private Class<?> forCommandClass = this.getClass();

	protected ACommand() {
		List<ISubCommandFactory> factories = SUB_COMMAND_FACTORIES.get(forCommandClass);
		if (factories != null) {
			MORE = new Object[factories.size()];
		} else {
			MORE = new Object[0];
		}
	}

	@Override
	public String toString() {

		StringBuilder sb = new StringBuilder();
		sb.append(super.toString());
		sb.append("[");
		for (Field propertyInfo : this.getClass().getDeclaredFields()) {
			sb.append(propertyInfo.getName());
			sb.append(":");
			try {
				sb.append(propertyInfo.get(this));
			} catch (IllegalArgumentException | IllegalAccessException e) {
				e.printStackTrace();
			}
			sb.append(";");
		}
		sb.append("]");

		return sb.toString();
	}
	
	public abstract String getCommandName();
	
	public String serialize(AProtocol protocol) {
		return protocol.serialize(this);
	}
	
	public void deserializeMore(String serializedMore, AProtocol protocol) {	
		Array array = protocol.deserializeArray(serializedMore, IInnerObject.class);
		Object[] more = array.values;
		List<ISubCommandFactory> factories = SUB_COMMAND_FACTORIES.get(forCommandClass);
		if (factories != null) {
			for (Object m : more) {
				for (ISubCommandFactory subCommandFactory : factories) {
					if (subCommandFactory.insert(m, MORE)) {
						break;
					}
				}
			}
		}
	}
}
